pin.go

```diff
@@ -251,3 +252,63 @@ func performPinChange(tpm *tpm2.TPMContext, public *tpm2.NVPublic, authPolicies
 
 	return nil
 }
+
+// ChangePIN changes the PIN for the key data file at the specified path. The existing PIN must be supplied via the oldPIN argument.
+// Setting newPIN to an empty string will clear the PIN and set a hint on the key data file that no PIN is set.
+//
+// If the TPM's dictionary attack logic has been triggered, a ErrTPMLockout error will be returned.
+//
+// If the file at the specified path cannot be opened, then a wrapped *os.PathError error will be returned.
+//
+// If the supplied key data file fails validation checks, an InvalidKeyFileError error will be returned.
+//
+// If oldPIN is incorrect, then a ErrPINFail error will be returned and the TPM's dictionary attack counter will be incremented.
+func ChangePIN(tpm *TPMConnection, path string, oldPIN, newPIN string) error {
+	// Check if the TPM is in lockout mode
+	props, err := tpm.GetCapabilityTPMProperties(tpm2.PropertyPermanent, 1)
+	if err != nil {
+		return xerrors.Errorf("cannot fetch properties from TPM: %w", err)
+	}
+
+	if tpm2.PermanentAttributes(props[0].Value)&tpm2.AttrInLockout > 0 {
+		return ErrTPMLockout
+	}
+
+	// Open the key data file
+	keyFile, err := os.Open(path)
+	if err != nil {
+		return xerrors.Errorf("cannot open key data file: %w", err)
+	}
+	defer keyFile.Close()
+
+	// Read and validate the key data file
+	data, _, pinIndexPublic, err := readAndValidateKeyData(tpm.TPMContext, keyFile, nil, tpm.HmacSession())
+	if err != nil {
+		var kfErr keyFileError
+		if xerrors.As(err, &kfErr) {
+			return InvalidKeyFileError{err.Error()}
+		}
+		return xerrors.Errorf("cannot read and validate key data file: %w", err)
+	}
+
+	// Change the PIN
+	if err := performPinChange(tpm.TPMContext, pinIndexPublic, data.StaticPolicyData.PinIndexAuthPolicies, oldPIN, newPIN, tpm.HmacSession()); err != nil {
+		if isAuthFailError(err, tpm2.CommandNVChangeAuth, 1) {
+			return ErrPINFail
+		}
+		return err
+	}
+
+	// Update the metadata and write a new key data file
```

Comment:
should we do the writing only if the hint has changed?

Reply:
That sounds ok to me.